question25/main.js

```diff
@@ -0,0 +1,114 @@
+'use strict';
+
+const textLinkToTerms = document.getElementById('js-terms-textlink');
+const modal = document.getElementById('js-modal');
+const modalClose = document.getElementById('js-close');
+const mask = document.getElementById('js-mask');
+const checkbox = document.getElementById('js-checkbox');
+const submitButton = document.getElementById('js-submit-button');
+const modalContainer = document.getElementById("js-modal-container");
+const nameInputArea = document.getElementById('name');
+const mailInputArea = document.getElementById('mail');
+const passwordInputArea = document.getElementById('password');
+
+
+textLinkToTerms.addEventListener('click',()=>{
+    modal.classList.remove('hidden');
+    mask.classList.remove('hidden');
+});
+modalClose.addEventListener('click',()=>{
+    modal.classList.add('hidden');
+    mask.classList.add('hidden');
+});
+mask.addEventListener('click',()=>{
+    modalClose.click();
+});
+
+const options = {
+  root: document.getElementById('js-modal'),
+  threshold: 1
+};
+
+const checkWhenIntersect = ([entry]) => {
+  checkbox.checked = entry.isIntersecting;
+  checkbox.disabled = !entry.isIntersecting;
+  entry.isIntersecting && checkAllValidity();
+};
+
+const observer = new IntersectionObserver(checkWhenIntersect, options);
+observer.observe(modalContainer.lastElementChild);
+
+
+
+const isValidStatus = {
+  name: false,
+  mail: false,
+  password: false,
+}
+
+const validationInfo = {
+  name: {
+    maxNameLength: 16,
+    minNameLength: 1,
+    validation: (value) => value.length > validationInfo.name.minNameLength && value.length < validationInfo.name.maxNameLength,
+    errorMessage: '※ユーザー名は1文字以上15文字以下にしてください。',
+  },
+  mail: {
+    //Reference: https://1-notes.com/javascript-determine-if-it-is-an-email-address/ 
+    validation : (value) => /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(value),
+    errorMessage: '※メールアドレスの形式になっていません。' 
+  },
+  password: {
+    validation : (value) => /(?=.*[a-zA-Z])(?=.*[0-9])[a-zA-z0-9]{8,}/.test(value),
+    errorMessage: '※8文字以上の大小の英数字を交ぜたものにしてください。'  
+  }
+}
+
+const setValidationEvents = (isValid, targetForm) => {
+  if (isValid) {
+    targetForm.classList.add('valid');
+    targetForm.classList.remove('invalid');
+    targetForm.nextElementSibling.textContent = ''; 
+    isValidStatus[targetForm.id] = true;
+  } else {
+    targetForm.classList.add('invalid');
+    targetForm.classList.remove('valid');
+    targetForm.nextElementSibling.textContent = validationInfo[targetForm.id].errorMessage;
+    isValidStatus[targetForm.id] = false; 
+  }
+}
+
+const validateInputValue = (e) => {
+  const targetForm = e.target;
+  const value = targetForm.value.trim();
+  const result = validationInfo[targetForm.id].validation(value);
+  setValidationEvents(result,targetForm);
+  checkEmptyCharacter(targetForm);  
+  checkAllValidity();
```

Comment:
Sorry. check** was used here too.
I refactored this code with my own interpretation, but what do you think?
```suggestion
  renderRequiredFieldMessages(targetForm);
  switchSubmitButton(checkAllValidity());
```

`checkEmptyCharacter`
The content of this function was to output error messages. So we kept it simple and made it this way.

`switchSubmitButton(checkAllValidity());`
I felt it was too much responsibility to have other functions executed in the function to be checked.
I was thinking that I could make it a function that simply returns a boolean value and put it as an argument to the switchSubmitButton.
```js
const checkAllValidity = () => {
  return Object.values(isValidStatus).every((result) => result);
};
```
At the event definition, you can execute it by passing an object as an argument.
```js
checkbox.addEventListener("input", {
  isValid: checkAllValidity,
  handleEvent: switchSubmitButton
});
```

[codesandbox](https://codesandbox.io/s/compassionate-hawking-ucnqqv?file=/main.js)

Reply:
Thank you for your review! 
I was actually a little uncomfortable with the inclusion of the switchSubmitButton in checkAllValidity. Thanks for the reference code...! I learned a lot!! I have corrected it accordingly :bow:

https://github.com/csakyo/homework/pull/39/commits/88ba15608d36311bfa9da26a6c317d51a4e4c58d

Reply:
@443n0511 
以前ご指摘いただいた箇所で、下記の関数のif文を省略したことで、動きが変わってしまっていたことに気づきました。(スクロールが下までいっても、上に戻るとチェックが外れてしまう)。entry.isIntersectingがtrueの時checkedだから、それ以外の時は外れてしまうのかな？と思ったのですが、細かい原因まではよくわかりませんでした...
if()で書くと問題なさそうなので、一旦元に戻しました。

修正後
```
const checkWhenIntersect = ([entry]) => {
  checkbox.checked = entry.isIntersecting;
  checkbox.disabled = !entry.isIntersecting;
  entry.isIntersecting && switchSubmitButton(checkAllValidity());
};
```

現在元に戻しました↓
```
const checkWhenIntersect = ([entry]) => {
  if (entry.isIntersecting) {
    checkbox.checked = true;
    checkbox.disabled = false;
    switchSubmitButton(checkAllValidity());
  };
};
```
https://github.com/csakyo/homework/pull/39/commits/88ba15608d36311bfa9da26a6c317d51a4e4c58d

Reply:
> 動きが変わってしまっていた
すみません！私の確認不足でした。修正ありがとうございます！